docs/architecture.md

```diff
@@ -8,16 +8,109 @@ This whole application was built off of [Node.js](https://nodejs.org/en/)!
 
 We use [Express.js](https://expressjs.com/) to help us with routing.
 
-### Authentication
+### Authentication library
 
 We use a `LocalStrategy` within Passport.js to authenticate a login session. Given an active session, the user must have a token called `session` in their cookies which was given by the api in order to continue accessing the api.
 
-In addition, authentication is currently being transitioned from some custom architecture to a more classic RBAC model, so that we are able to only allow access to certain routes given a user has a specific role.
+For authorization, we decided to write a custom RBAC-esque strategy. Check below for specifics on how it works, and how to work with it, and how to use it when extending this project.
 
 ### Testing
 
 We use [Mocha](https://mochajs.org/) with [Chai](https://www.chaijs.com/) to test our routes and services.
 
+## Authorization
+
+Before you read this section, it's important to understand the difference between `Authentication` and `Authorization`. The first asks the question: _"Do we know who you are?"_ (i.e. are you logged in?), whereas the second asks the question: _"Do you have sufficient permissions to be making this request?"_.
+
+We use a custom implementation for authorization, which relies heavily on the design of Role-based access control.
+
+### Definitions
+
+#### `Route`
+
+A `Route` is defined by its uri path, its query parameters in the uri path, and the HTTP verb used to access it:
+
+```json
+{
+    "requestType": "POST",
+    "uri": "/api/sponsor/",
+}
+```
+
+#### `Role`
+
+A `Role` is a collection of `Routes`, and a unique name (such as `hacker`, or `sponsor`, or `my-custom-role`). This collection of `Routes` should approximately encapsulate the set of actions that a user of the api should be able to do. For example, a `Hacker` might want to have access to `GET /api/hacker/507f1f77bcf86cd799439011` (get their own information), or `POST /api/hacker/` (create their hacker document):
+
+```json
+{
+    "accountId": "507f191e810c19729de860ea",
+    "roles": ["54759eb3c090d83494e2d804", "51325eb3c090d83494e2d702"],
+}
```

Comment:
This JSON object follows the model of RoleBinding. Perhaps you meant to put it below?
The Role model is:
`{Name: string, routes: Array of {uri: string, requestType: {string enum of request types}}}`
I think it's also important to note that the we do not have the concept of 'Routes' in our schema. I'm only using it in the tests to make creating roles easier, as there are some routes that are used in multiple roles.

Reply:
For our own testing, routes are grouped by the type in /api/type/... these routes will have a key, and the name of the singular role will be the key name plus the type. Ex: getSelfByIdHacker.

Reply:
Also it might be important to note the use case of a custom rolebinding for an account. In the case that an individual has some baseline role + some extra permissions, we are adding singular roles to the rolebinding, instead of creating a new role. This means that there needs to be a role for each route. We may want to add this information to the docs

Reply:
Yeah, I'll write documentation about what role bindings we've created once we actually have the .js file set up for the database creation.